libcontainer/cgroups/fs/devices.go

```diff
@@ -32,29 +40,63 @@ func (s *DevicesGroup) Set(path string, cgroup *configs.Cgroup) error {
 	}
 
 	devices := cgroup.Resources.Devices
+	oldAllowedDevices, err := readDevicesExceptDefault(path)
+	if err != nil {
+		return err
+	}
+
 	if len(devices) > 0 {
 		for _, dev := range devices {
 			file := "devices.deny"
 			if dev.Allow {
 				file = "devices.allow"
 			}
-			if err := fscommon.WriteFile(path, file, dev.CgroupString()); err != nil {
-				return err
+
+			// For the second time set, we don't deny all devices
+			if dev.Type == defaultDevice.Type && len(oldAllowedDevices) != 0 {
+				file = ""
+			}
+
+			if len(file) > 0 {
+				if err := fscommon.WriteFile(path, file, dev.CgroupString()); err != nil {
+					return err
+				}
 			}
 		}
 		return nil
 	}
 	if cgroup.Resources.AllowAllDevices != nil {
 		if *cgroup.Resources.AllowAllDevices == false {
```

Comment:
This isn't directly related to this patch (and is something we should fix separately), but I've hit this wall while trying to review it.

I've spent a while now trying to understand what the hell this whole `AllowAllDevices`, `AllowedDevices` and `DeniedDevices` stuff is supposed to do and how it interacts with `Devices`. It's been around forever in runc, isn't exposed in the runtime-spec, and isn't even generated by specconv. It appears to be only used for unit testing itself, which seems like a senseless thing to test. Honestly, this patch and many others would be much nicer if we just nuked this from orbit and just had `Devices`.

In the context of this patch, I'm not even sure how to review whether the semantics of the below code are correct -- if I'm correct in my understanding that nothing uses the code, how do we know if the changes make sense...

@crosbymichael Am I missing something? Will something really bad happen if we get rid of it? I remember the `initProcessTime` nightmare when we changed the type of a field in the libcontainer state JSON (which resulted in the whole "migration" fun) -- would something similar happen if we just purged this concept from the codebase?

@AkihiroSuda WDYT? (Since you were the last person to touch it, and I think @crosbymichael and I last looked at this code in 2016.)